File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDurationExtractor.cs
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Linq;
using Microsoft.Recognizers.Text.Utilities;
using DateObject = System.DateTime;

namespace Microsoft.Recognizers.Text.DateTime
{
    public class BaseDurationExtractor : IDateTimeExtractor
    {
        public static readonly string ExtractorName = Constants.SYS_DATETIME_DURATION;

        private readonly IDurationExtractorConfiguration config;

        private readonly bool merge;

        public BaseDurationExtractor(IDurationExtractorConfiguration config, bool merge = true)
        {
            this.config = config;
            this.merge = merge;
        }

        public List<ExtractResult> Extract(string text)
        {
            return Extract(text, DateObject.Now);
        }

        public List<ExtractResult> Extract(string text, DateObject reference)
        {
            var tokens = new List<Token>();
            var numberWithUnitTokens = NumberWithUnit(text);

            tokens.AddRange(numberWithUnitTokens);
            tokens.AddRange(NumberWithUnitAndSuffix(text, numberWithUnitTokens));
            tokens.AddRange(ImplicitDuration(text));

            var rets = Token.MergeAllTokens(tokens, text, ExtractorName);

            // First MergeMultipleDuration then ResolveMoreThanOrLessThanPrefix so cases like "more than 4 days and less than 1 week" will not be merged into one "multipleDuration"
            if (this.merge)
            {
                rets = MergeMultipleDuration(text, rets);
            }

            rets = TagInequalityPrefix(text, rets);

            return rets;
        }

        // handle cases look like: {more than | less than} {duration}?
        private List<ExtractResult> TagInequalityPrefix(string text, List<ExtractResult> ers)
        {
            foreach (var er in ers)
            {
                var beforeString = text.Substring(0, (int)er.Start);
                var afterString = text.Substring((int)er.Start + (int)er.Length);
                bool isInequalityPrefixMatched = false;
                bool isMatchAfter = false;

                var match = config.MoreThanRegex.MatchEnd(beforeString, trim: true);

                // check also afterString
                if (this.config.CheckBothBeforeAfter && !match.Success)
                {
                    match = config.MoreThanRegex.MatchesBegin(afterString, trim: true);
                    isMatchAfter = true;
                }

                // The second condition is necessary so for "1 week" in "more than 4 days and less than 1 week", it will not be tagged incorrectly as "more than"
                if (match.Success)
                {
                    er.Data = Constants.MORE_THAN_MOD;
                    isInequalityPrefixMatched = true;
                }

                if (!isInequalityPrefixMatched)
                {
                    match = config.LessThanRegex.MatchEnd(beforeString, trim: true);

                    // check also afterString
                    if (this.config.CheckBothBeforeAfter && !match.Success)
                    {
                        match = config.LessThanRegex.MatchesBegin(afterString, trim: true);
                        isMatchAfter = true;
                    }

                    if (match.Success)
                    {
                        er.Data = Constants.LESS_THAN_MOD;
                        isInequalityPrefixMatched = true;
                    }
                }

                if (isInequalityPrefixMatched)
                {
                    if (!isMatchAfter)
                    {
                        er.Length += er.Start - match.Index;
                        er.Start = match.Index;
                        er.Text = text.Substring((int)er.Start, (int)er.Length);
                    }
                    else
                    {
                        er.Length += match.Index + match.Length;
                        er.Text = text.Substring((int)er.Start, (int)er.Length);
                    }
                }
            }

            return ers;
        }

        // handle cases look like: {number} {unit}? and {an|a} {half|quarter} {unit}?
        // define the part "and {an|a} {half|quarter}" as Suffix
        private List<Token> NumberWithUnitAndSuffix(string text, List<Token> ers)
        {
            var ret = new List<Token>();
            foreach (var er in ers)
            {
                var afterStr = text.Substring(er.Start + er.Length);
                var match = this.config.SuffixAndRegex.MatchBegin(afterStr, trim: true);

                if (match.Success)
                {
                    ret.Add(new Token(er.Start, (er.Start + er.Length) + match.Length));
                }
                else if (this.config.CheckBothBeforeAfter)
                {
                    // check also beforeStr
                    var beforeStr = text.Substring(0, er.Start);
                    match = this.config.SuffixAndRegex.MatchEnd(beforeStr, trim: true);
                    if (match.Success)
                    {
                        ret.Add(new Token(match.Index, er.Start + er.Length));
                    }
                }
            }

            return ret;
        }

        // simple cases of a number followed by unit
        private List<Token> NumberWithUnit(string text)
        {
            var ret = new List<Token>();
            var ers = ExtractNumbersBeforeUnit(text);

            foreach (var er in ers)
            {
                var afterStr = text.Substring(er.Start + er.Length ?? 0);
                var match = this.config.FollowedUnit.MatchBegin(afterStr, trim: true);

                if (match.Success)
                {
                    ret.Add(new Token(er.Start ?? 0, (er.Start + er.Length ?? 0) + match.Length));
                }
            }

            // handle "3hrs"
            ret.AddRange(Token.GetTokenFromRegex(config.NumberCombinedWithUnit, text));

            // handle "an hour"
            ret.AddRange(Token.GetTokenFromRegex(config.AnUnitRegex, text));

            // handle "few" related cases
            ret.AddRange(Token.GetTokenFromRegex(config.InexactNumberUnitRegex, text));

            return ret;
        }

        // @TODO improve re-use with Parser
        private List<ExtractResult> ExtractNumbersBeforeUnit(string text)
        {
            var ers = this.config.CardinalExtractor.Extract(text);

            // In special cases some languages will treat "both" as a number to be combined with duration units.
            var specialNumberUnitTokens = Token.GetTokenFromRegex(config.SpecialNumberUnitRegex, text);

            foreach (var token in specialNumberUnitTokens)
            {
                var er = new ExtractResult
                {
                    Start = token.Start,
                    Length = token.Length,
                    Text = text.Substring(token.Start, token.Length),
                };

                ers.Add(er);
            }

            return ers;
        }

        // handle cases that don't contain number
        private List<Token> ImplicitDuration(string text)
        {
            var ret = new List<Token>();

            // handle "all day", "all year"
            ret.AddRange(Token.GetTokenFromRegex(config.AllRegex, text));

            // handle "half day", "half year"
            ret.AddRange(Token.GetTokenFromRegex(config.HalfRegex, text));

            // handle "next day", "last year"
            ret.AddRange(Token.GetTokenFromRegex(config.RelativeDurationUnitRegex, text));

            // handle "during/for the day/week/month/year"
            if ((config.Options & DateTimeOptions.CalendarMode) != 0)
            {
                ret.AddRange(Token.GetTokenFromRegex(config.DuringRegex, text));
            }

            return ret;
        }

        private List<ExtractResult> MergeMultipleDuration(string text, List<ExtractResult> extractorResults)
        {
            if (extractorResults.Count <= 1)
            {
                return extractorResults;
            }

            var unitMap = this.config.UnitMap;
            var unitValueMap = this.config.UnitValueMap;
            var unitRegex = this.config.DurationUnitRegex;
            List<ExtractResult> results = new List<ExtractResult>();
            List<List<ExtractResult>> separateResults = new List<List<ExtractResult>>();

            var firstExtractionIndex = 0;
            var timeUnit = 0;
            var totalUnit = 0;
            while (firstExtractionIndex < extractorResults.Count)
            {
                string curUnit = null;
                var unitMatch = unitRegex.Match(extractorResults[firstExtractionIndex].Text);

                if (unitMatch.Success && unitMap.ContainsKey(unitMatch.Groups["unit"].ToString()))
                {
                    curUnit = unitMatch.Groups["unit"].ToString();
                    totalUnit++;
                    if (DurationParsingUtil.IsTimeDurationUnit(unitMap[curUnit]))
                    {
                        timeUnit++;
                    }
                }

                if (string.IsNullOrEmpty(curUnit))
                {
                    firstExtractionIndex++;
                    continue;
                }

                // Add extraction to list of separate results (needed in case the extractions should not be merged)
                List<ExtractResult> separateList = new List<ExtractResult>() { extractorResults[firstExtractionIndex] };

                var secondExtractionIndex = firstExtractionIndex + 1;
                while (secondExtractionIndex < extractorResults.Count)
                {
                    var valid = false;
                    var midStrBegin = extractorResults[secondExtractionIndex - 1].Start + extractorResults[secondExtractionIndex - 1].Length ?? 0;
                    var midStrEnd = extractorResults[secondExtractionIndex].Start ?? 0;
                    var midStr = text.Substring(midStrBegin, midStrEnd - midStrBegin);
                    var match = this.config.DurationConnectorRegex.Match(midStr);
                    if (match.Success)
                    {
                        unitMatch = unitRegex.Match(extractorResults[secondExtractionIndex].Text);
                        if (unitMatch.Success && unitMap.ContainsKey(unitMatch.Groups["unit"].ToString()))
                        {
                            var nextUnitStr = unitMatch.Groups["unit"].ToString();
                            if (unitValueMap[nextUnitStr] != unitValueMap[curUnit])
                            {
                                valid = true;
                                if (unitValueMap[nextUnitStr] < unitValueMap[curUnit])
                                {
                                    curUnit = nextUnitStr;
                                }
                            }

                            totalUnit++;
                            if (DurationParsingUtil.IsTimeDurationUnit(unitMap[nextUnitStr]))
                            {
                                timeUnit++;
                            }
                        }
                    }

                    if (!valid)
                    {
                        break;
                    }

                    // Add extraction to list of separate results (needed in case the extractions should not be merged)
                    separateList.Add(extractorResults[secondExtractionIndex]);

                    secondExtractionIndex++;
                }

                if (secondExtractionIndex - 1 > firstExtractionIndex)
                {
                    var node = new ExtractResult();
                    node.Start = extractorResults[firstExtractionIndex].Start;
                    node.Length = extractorResults[secondExtractionIndex - 1].Start + extractorResults[secondExtractionIndex - 1].Length - node.Start;
                    node.Text = text.Substring(node.Start ?? 0, node.Length ?? 0);
                    node.Type = extractorResults[firstExtractionIndex].Type;

                    // Add multiple duration type to extract result
                    string type = Constants.MultipleDuration_DateTime; // Default type
                    if (timeUnit == totalUnit)
                    {
                        type = Constants.MultipleDuration_Time;
                    }
                    else if (timeUnit == 0)
                    {
                        type = Constants.MultipleDuration_Date;
                    }

                    node.Data = type;

                    results.Add(node);

                    timeUnit = 0;
                    totalUnit = 0;
                }
                else
                {
                    results.Add(extractorResults[firstExtractionIndex]);
                }

                // Add list of separate extractions to separateResults, so that there is a 1 to 1 correspondence
                // between results (list of merged extractions) and separateResults (list of unmerged extractions)
                separateResults.Add(separateList);

                firstExtractionIndex = secondExtractionIndex;
            }

            // If the first and last elements of a group of contiguous extractions are both preceded/followed by modifiers,
            // they should not be merged, e.g. "last 2 weeks and 3 days ago"
            for (int i = results.Count - 1; i >= 0; i--)
            {
                var start = (int)results[i].Start;
                var end = start + (int)results[i].Length;
                var beforeStr = text.Substring(0, start);
                var afterStr = text.Substring(end);
                var beforeMod = this.config.ModPrefixRegex.MatchEnd(beforeStr, trim: true);
                var afterMod = this.config.ModSuffixRegex.MatchBegin(afterStr, trim: true);
                if (beforeMod.Success && afterMod.Success)
                {
                    results.RemoveAt(i);
                    results.InsertRange(i, separateResults[i]);
                }
            }

            return results;
        }
    }
}

File: Java/libraries/recognizers-text-date-time/src/main/java/com/microsoft/recognizers/text/datetime/extractors/BaseDurationExtractor.java
package com.microsoft.recognizers.text.datetime.extractors;

import com.google.common.collect.ImmutableMap;
import com.microsoft.recognizers.text.ExtractResult;
import com.microsoft.recognizers.text.datetime.Constants;
import com.microsoft.recognizers.text.datetime.DateTimeOptions;
import com.microsoft.recognizers.text.datetime.extractors.config.IDurationExtractorConfiguration;
import com.microsoft.recognizers.text.datetime.utilities.ConditionalMatch;
import com.microsoft.recognizers.text.datetime.utilities.DurationParsingUtil;
import com.microsoft.recognizers.text.datetime.utilities.RegexExtension;
import com.microsoft.recognizers.text.datetime.utilities.Token;
import com.microsoft.recognizers.text.utilities.Match;
import com.microsoft.recognizers.text.utilities.RegExpUtility;
import com.microsoft.recognizers.text.utilities.StringUtility;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class BaseDurationExtractor implements IDateTimeExtractor {

    private final IDurationExtractorConfiguration config;
    private final boolean merge;

    @Override
    public String getExtractorName() {
        return Constants.SYS_DATETIME_DURATION;
    }

    public BaseDurationExtractor(IDurationExtractorConfiguration config) {
        this(config, true);
    }

    public BaseDurationExtractor(IDurationExtractorConfiguration config, boolean merge) {
        this.config = config;
        this.merge = merge;
    }

    @Override
    public List<ExtractResult> extract(String input) {
        return this.extract(input, LocalDateTime.now());
    }

    @Override
    public List<ExtractResult> extract(String input, LocalDateTime reference) {
        List<Token> tokens = new ArrayList<>();

        List<Token> numberWithUnitTokens = numberWithUnit(input);

        tokens.addAll(numberWithUnitTokens);
        tokens.addAll(numberWithUnitAndSuffix(input, numberWithUnitTokens));
        tokens.addAll(implicitDuration(input));

        List<ExtractResult> result = Token.mergeAllTokens(tokens, input, getExtractorName());

        // First MergeMultipleDuration then ResolveMoreThanOrLessThanPrefix so cases like "more than 4 days and less than 1 week" will not be merged into one "multipleDuration"
        if (merge) {
            result = mergeMultipleDuration(input, result);
        }

        result = tagInequalityPrefix(input, result);

        return result;
    }

    private List<ExtractResult> tagInequalityPrefix(String input, List<ExtractResult> result) {
        Stream<ExtractResult> resultStream = result.stream().map(er -> {
            String beforeString = input.substring(0, er.getStart());
            boolean isInequalityPrefixMatched = false;

            ConditionalMatch match = RegexExtension.matchEnd(this.config.getMoreThanRegex(), beforeString, true);

            // The second condition is necessary so for "1 week" in "more than 4 days and less than 1 week", it will not be tagged incorrectly as "more than"
            if (match.getSuccess()) {
                er.setData(Constants.MORE_THAN_MOD);
                isInequalityPrefixMatched = true;
            }

            if (!isInequalityPrefixMatched) {
                match = RegexExtension.matchEnd(this.config.getLessThanRegex(), beforeString, true);

                if (match.getSuccess()) {
                    er.setData(Constants.LESS_THAN_MOD);
                    isInequalityPrefixMatched = true;
                }
            }

            if (isInequalityPrefixMatched) {
                int length = er.getLength() + er.getStart() - match.getMatch().get().index;
                int start = match.getMatch().get().index;
                String text = input.substring(start, start + length);
                er.setStart(start);
                er.setLength(length);
                er.setText(text);
            }

            return er;
        });
        return resultStream.collect(Collectors.toList());
    }

    private List<ExtractResult> mergeMultipleDuration(String input, List<ExtractResult> extractResults) {
        if (extractResults.size() <= 1) {
            return extractResults;
        }

        ImmutableMap<String, String> unitMap = config.getUnitMap();
        ImmutableMap<String, Long> unitValueMap = config.getUnitValueMap();
        Pattern unitRegex = config.getDurationUnitRegex();

        List<ExtractResult> result = new ArrayList<>();

        int firstExtractionIndex = 0;
        int timeUnit = 0;
        int totalUnit = 0;

        while (firstExtractionIndex < extractResults.size()) {
            String currentUnit = null;
            Optional<Match> unitMatch = Arrays.stream(RegExpUtility.getMatches(unitRegex, extractResults.get(firstExtractionIndex).getText())).findFirst();

            if (unitMatch.isPresent() && unitMap.containsKey(unitMatch.get().getGroup("unit").value)) {
                currentUnit = unitMatch.get().getGroup("unit").value;
                totalUnit++;
                if (DurationParsingUtil.isTimeDurationUnit(unitMap.get(currentUnit))) {
                    timeUnit++;
                }
            }

            if (StringUtility.isNullOrEmpty(currentUnit)) {
                firstExtractionIndex++;
                continue;
            }

            int secondExtractionIndex = firstExtractionIndex + 1;
            while (secondExtractionIndex < extractResults.size()) {
                boolean valid = false;
                int midStrBegin = extractResults.get(secondExtractionIndex - 1).getStart() + extractResults.get(secondExtractionIndex - 1).getLength();
                int midStrEnd = extractResults.get(secondExtractionIndex).getStart();
                String midStr = input.substring(midStrBegin, midStrEnd);
                Optional<Match> match = Arrays.stream(RegExpUtility.getMatches(this.config.getDurationConnectorRegex(), midStr)).findFirst();

                if (match.isPresent()) {
                    unitMatch = Arrays.stream(RegExpUtility.getMatches(unitRegex, extractResults.get(secondExtractionIndex).getText())).findFirst();

                    if (unitMatch.isPresent() && unitMap.containsKey(unitMatch.get().getGroup("unit").value)) {
                        String nextUnitStr = unitMatch.get().getGroup("unit").value;

                        if (unitValueMap.get(nextUnitStr) != unitValueMap.get(currentUnit)) {
                            valid = true;

                            if (unitValueMap.get(nextUnitStr) < unitValueMap.get(currentUnit)) {
                                currentUnit = nextUnitStr;
                            }
                        }

                        totalUnit++;

                        if (DurationParsingUtil.isTimeDurationUnit(unitMap.get(nextUnitStr))) {
                            timeUnit++;
                        }
                    }
                }

                if (!valid) {
                    break;
                }

                secondExtractionIndex++;
            }

            if (secondExtractionIndex - 1 > firstExtractionIndex) {
                int start = extractResults.get(firstExtractionIndex).getStart();
                int length = extractResults.get(secondExtractionIndex - 1).getStart() + extractResults.get(secondExtractionIndex - 1).getLength() - start;
                String text = input.substring(start, start + length);
                String rType = extractResults.get(firstExtractionIndex).getType();
                ExtractResult node = new ExtractResult(start, length, text, rType, null);

                // add multiple duration type to extract result
                String type = null;

                if (timeUnit == totalUnit) {
                    type = Constants.MultipleDuration_Time;
                } else if (timeUnit == 0) {
                    type = Constants.MultipleDuration_Date;
                } else {
                    type = Constants.MultipleDuration_DateTime;
                }

                node.setData(type);
                result.add(node);

                timeUnit = 0;
                totalUnit = 0;

            } else {
                result.add(extractResults.get(firstExtractionIndex));
            }

            firstExtractionIndex = secondExtractionIndex;
        }

        return result;
    }

    // handle cases that don't contain nubmer
    private Collection<Token> implicitDuration(String text) {
        Collection<Token> result = new ArrayList<>();

        // handle "all day", "all year"
        result.addAll(getTokenFromRegex(config.getAllRegex(), text));

        // handle "half day", "half year"
        result.addAll(getTokenFromRegex(config.getHalfRegex(), text));

        // handle "next day", "last year"
        result.addAll(getTokenFromRegex(config.getRelativeDurationUnitRegex(), text));

        // handle "during/for the day/week/month/year"
        if (config.getOptions().match(DateTimeOptions.CalendarMode)) {
            result.addAll(getTokenFromRegex(config.getDuringRegex(), text));
        }

        return result;
    }

    // simple cases made by a number followed an unit
    private List<Token> numberWithUnit(String text) {
        List<Token> result = new ArrayList<>();
        List<ExtractResult> ers = this.config.getCardinalExtractor().extract(text);
        for (ExtractResult er : ers) {
            String afterStr = text.substring(er.getStart() + er.getLength());
            ConditionalMatch match = RegexExtension.matchBegin(this.config.getFollowedUnit(), afterStr, true);
            if (match.getSuccess() && match.getMatch().get().index == 0) {
                result.add(new Token(er.getStart(), er.getStart() + er.getLength() + match.getMatch().get().length));
            }
        }

        // handle "3hrs"
        result.addAll(this.getTokenFromRegex(this.config.getNumberCombinedWithUnit(), text));

        // handle "an hour"
        result.addAll(this.getTokenFromRegex(this.config.getAnUnitRegex(), text));

        // handle "few" related cases
        result.addAll(this.getTokenFromRegex(this.config.getInexactNumberUnitRegex(), text));

        return result;
    }

    private Collection<Token> getTokenFromRegex(Pattern pattern, String text) {
        Collection<Token> result = new ArrayList<>();

        for (Match match : RegExpUtility.getMatches(pattern, text)) {
            result.add(new Token(match.index, match.index + match.length));
        }

        return result;
    }

    // handle cases look like: {number} {unit}? and {an|a} {half|quarter} {unit}?
    // define the part "and {an|a} {half|quarter}" as Suffix
    private Collection<Token> numberWithUnitAndSuffix(String text, Collection<Token> tokens) {
        Collection<Token> result = new ArrayList<>();

        for (Token token : tokens) {
            String afterStr = text.substring(token.getStart() + token.getLength());
            ConditionalMatch match = RegexExtension.matchBegin(this.config.getSuffixAndRegex(), afterStr, true);
            if (match.getSuccess() && match.getMatch().get().index == 0) {
                result.add(new Token(token.getStart(), token.getStart() + token.getLength() + match.getMatch().get().length));
            }
        }

        return result;
    }
}
